### PR TITLE
[Merged by Bors] - feat: general Ascoli theorem

### PR DESCRIPTION
**Pull request by ADedecker**
Co-authored-by: @vbeffara (port to Lean4), @tb65536 (suggested to skip the totally bounded case and go straight to compactness using Tykhonov)

This was discussed on Zulip [recently](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Arzela-Ascoli.20for.20uniform.20spaces) and [a while ago](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Montel's.20theorem).

---

- [x] depends on: #9286
- [x] depends on: #10143
- [x] depends on: #10145
- [x] depends on: #10153 
- [x] depends on: #10161 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by tb65536** [comment]
Does this allow you to golf the existing Arzela-Ascoli theorem for metric spaces?

**Comment by vbeffara** [comment]
I removed 3 lemmas that were probably duplicated by a previous merge.

**Comment by ADedecker** [comment]
Thanks Vincent! Thomas, yes that should be quite easy but let's keep that for a later PR.

Sorry to both of you, the beginning of the year has been a bit more chaotic than expected but I should have more time rather soon.

File: Mathlib/Topology/UniformSpace/Ascoli.lean
Comment (ocfnash) [inline Mathlib/Topology/UniformSpace/Ascoli.lean:46]:
```suggestion
  is strictly weaker.
```

File: Mathlib/Topology/UniformSpace/Ascoli.lean
Comment (ocfnash) [inline Mathlib/Topology/UniformSpace/Ascoli.lean:50]: Could you clarify what you mean by "this" here: I presume you mean establishing the intermediate total boundedness result a la Bourbaki?

File: Mathlib/Topology/UniformSpace/Ascoli.lean
Comment (ocfnash) [inline Mathlib/Topology/UniformSpace/Ascoli.lean:71]:
```suggestion
open Set Filter Uniformity Topology Function UniformConvergence
```

File: Mathlib/Topology/UniformSpace/Ascoli.lean
Comment (ocfnash) [inline Mathlib/Topology/UniformSpace/Ascoli.lean:73]:
```suggestion
variable {ι X Y α β : Type*} [TopologicalSpace X] [UniformSpace α] [UniformSpace β]
```

**Review by ocfnash — APPROVED** [review]
This is awesome! My only regret is that I could not come up with any non-trivial suggestions while reviewing.

bors d+

File: Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean
Comment (ADedecker) [reply Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean:883]: I agree with you to some extent, although working with beginners at LFTCM made me remember that it takes some work to be able to *read* Lean statements. I'll admit that sometimes I do so just because it takes more energy to write down the full statement without referencing the Lean objects at play.

In this particular case though, I think it looks useless in the source but not so much in the generated documentation because of the lack of type ascriptions. For example you can look at the similar [UniformOnFun.uniformContinuous_restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/UniformConvergenceTopology.html#UniformOnFun.uniformContinuous_restrict), where the docstring improves readability just by making the types explicit.

Comment (ADedecker) [reply Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean:883]: Although now that I say that, the statement I linked to doesn't write down the Lean term for the restriction map. Should I do the same here?

File: Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean
Comment (ADedecker) [reply Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean:620]: Indeed, I think this is what happened. Reverted.

File: Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean
Comment (ADedecker) [reply Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean]: I did that for now, let me know what you think.

Comment (ocfnash) [reply Mathlib/Topology/UniformSpace/UniformConvergenceTopology.lean]: What you've got is great, I don't think we need to hold up this really useful PR any more by my quibbles.

**Comment by ocfnash** [comment]
bors merge

File: Mathlib/Topology/UniformSpace/Ascoli.lean
Comment (ADedecker) [reply Mathlib/Topology/UniformSpace/Ascoli.lean:49]: Don't you want me to change this before merging? I was planning to do it tomorrow morning. I can also change the wording later if you want.

Comment (ocfnash) [reply Mathlib/Topology/UniformSpace/Ascoli.lean:49]: I decided I was splitting hairs and it wasn't worth delaying this PR any further.

Feel free to follow up with a doc string adjustment PR or not, according to your taste!

**Pull request closed without merging**
